### PR TITLE
Real-time message processing and synchronized database updates

### PR DESCRIPTION
**Pull request by nateyazdani**
I've written a working implementation of my second proposed re-architecturing solution, described in my issue [here](https://github.com/fedora-infra/statscache/issues/23#issuecomment-126097527). During normal operation, this is not that big of a change. When a message is received, a deep copy is directly handed to each plugin by the fedmsg consumer for processing, and each plugin stores whatever the results are of this processing internally. For example, volume plugins do something along the lines of `self._volumes[(msg.timestamp, key(msg))]  += 1` where `self._volumes` is a `collections.defaultdict`. On a configurable frequency, a polling producer instructs each plugin to update its database model with its stored results.

At startup, things do change quite a bit. For one, there is no longer a need for such a complicated plugin/producer initialization relationship (where the polling producers needed to be generated after plugins were loaded but before they were instantiated). Additionally, the plugin `initialize()` method, used to process backlogged messages, is removed in favor of doing this centrally in the fedmsg consumer initialization. This is good for encapsulation, as all operations related to message retrieval, receipt, and distribution happen in the same class. Since patchy data can't give us all that much insight, I've also added a configuration value for the "epoch" of statistics. This is the date and time as of which statscache will have uninterruptedly processed all fedmsg traffic. However, some plugins (the releng one in particular comes to mind) really do only care about somewhat recent messages. To support this, plugins are able to modify this continuity, basically by lying about how up-to-date they are (_i.e._, reporting that they are up-to-date as of `max(datetime.now() - self.backlog_delta, latest_row.timestamp)`). Of course, it must then be noted that historical data in such models will be spotty or possibly even nonexistent.

Note: Due to the fact that consensus was never reached in the original issue, please feel free to consider this an RFC or WIP if there are objections.

### CODE REVIEW

File: statscache/producer.py [hunk 1]
```diff
@@ -12,95 +11,35 @@
 log = logging.getLogger("fedmsg")
 
 
-class StatsProducerBase(moksha.hub.api.PollingProducer):
+class StatsProducer(moksha.hub.api.PollingProducer):
     """
     An abstract base class for dynamically generated producers. Subclasses need
```
Comment (ralphbean) [inline statscache/producer.py]: It seems this docstring is out of date, or am I misunderstanding things?  There is now only one polling producer in play, so there are no more dynamically generated producers.

**Comment by ralphbean** [comment]
This looks generally good.  I left some questions in fedora-infra/statscache_plugins#9.  :)

File: statscache/producer.py (hunk 1 above)
Comment (nateyazdani) [inline statscache/producer.py]: Oh no you're right; that definitely is out-of-date. Nice catch!

**Comment by nateyazdani** [comment]
Awesome, appreciate the review of all this code. I'll incorporate your suggestions and push the changes shortly.

**Comment by ralphbean** [comment]
:+1: here, as in fedora-infra/statscache_plugins#9

**Comment by nateyazdani** [comment]
Great, merging as well!

**Pull request merged**